From patchwork Thu May  7 18:56:13 2020
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: "Gustavo A. R. Silva" <gustavoars@kernel.org>
X-Patchwork-Id: 11534713
Return-Path: <SRS0=eykK=6V=vger.kernel.org=kvm-owner@kernel.org>
Received: from mail.kernel.org (pdx-korg-mail-1.web.codeaurora.org
 [172.30.200.123])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id ADD7F14B4
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu,  7 May 2020 18:51:51 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 972212495D
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu,  7 May 2020 18:51:51 +0000 (UTC)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/simple; d=kernel.org;
	s=default; t=1588877511;
	bh=+83EaeRY/7OLMLTOiUPtLBN9CV5OHRJu2uYjreh/MG0=;
	h=Date:From:To:Cc:Subject:List-ID:From;
	b=UgGtM+bh4zRHakd1r5o9/dB0n08EiTmegzgZOL1CuXnxxPuQyIULO5rg1tPN6yhna
	 lt8YTtUQ68mP1ZQoynoIAxgsbzg9v7NAy22jO88gBNHB6dwno5tHX+46UT7oVRFmWT
	 avX72i7TQmkBqAheIZPCQm/pm1v2lrDpBjPFzhis=
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1728743AbgEGSvr (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Thu, 7 May 2020 14:51:47 -0400
Received: from mail.kernel.org ([198.145.29.99]:52510 "EHLO mail.kernel.org"
        rhost-flags-OK-OK-OK-OK) by vger.kernel.org with ESMTP
        id S1728110AbgEGSvr (ORCPT <rfc822;kvm@vger.kernel.org>);
        Thu, 7 May 2020 14:51:47 -0400
Received: from embeddedor (unknown [189.207.59.248])
        (using TLSv1.2 with cipher ECDHE-RSA-AES256-GCM-SHA384 (256/256 bits))
        (No client certificate requested)
        by mail.kernel.org (Postfix) with ESMTPSA id E9A0624957;
        Thu,  7 May 2020 18:51:46 +0000 (UTC)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/simple; d=kernel.org;
        s=default; t=1588877507;
        bh=+83EaeRY/7OLMLTOiUPtLBN9CV5OHRJu2uYjreh/MG0=;
        h=Date:From:To:Cc:Subject:From;
        b=PJ9CpzulMqFvltlGIQj6v0A68/uG35qnUgsPfyAL/pyESRGYYy8Yvjecbm4OdF+mQ
         2E7OIngP0zMiAMNU7IMfUW3f891p4uTzeWROnLP9rhleYm3gYfEKPKK+r35gjqTpfm
         9BxPdRbQdLG3xoxDPiKXSq8Ay3jpaBIhZ/DHn5j4=
Date: Thu, 7 May 2020 13:56:13 -0500
From: "Gustavo A. R. Silva" <gustavoars@kernel.org>
To: Paolo Bonzini <pbonzini@redhat.com>
Cc: kvm@vger.kernel.org, linux-kernel@vger.kernel.org
Subject: [PATCH] KVM: Replace zero-length array with flexible-array
Message-ID: <20200507185613.GA14808@embeddedor>
MIME-Version: 1.0
Content-Disposition: inline
User-Agent: Mutt/1.9.4 (2018-02-28)
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

The current codebase makes use of the zero-length array language
extension to the C90 standard, but the preferred mechanism to declare
variable-length types such as these ones is a flexible array member[1][2],
introduced in C99:

struct foo {
        int stuff;
        struct boo array[];
};

By making use of the mechanism above, we will get a compiler warning
in case the flexible array does not occur last in the structure, which
will help us prevent some kind of undefined behavior bugs from being
inadvertently introduced[3] to the codebase from now on.

Also, notice that, dynamic memory allocations won't be affected by
this change:

"Flexible array members have incomplete type, and so the sizeof operator
may not be applied. As a quirk of the original implementation of
zero-length arrays, sizeof evaluates to zero."[1]

sizeof(flexible-array-member) triggers a warning because flexible array
members have incomplete type[1]. There are some instances of code in
which the sizeof operator is being incorrectly/erroneously applied to
zero-length arrays and the result is zero. Such instances may be hiding
some bugs. So, this work (flexible-array member conversions) will also
help to get completely rid of those sorts of issues.

This issue was found with the help of Coccinelle.

[1] https://gcc.gnu.org/onlinedocs/gcc/Zero-Length.html
[2] https://github.com/KSPP/linux/issues/21
[3] commit 76497732932f ("cxgb3/l2t: Fix undefined behaviour")

Signed-off-by: Gustavo A. R. Silva <gustavoars@kernel.org>
---
 include/linux/kvm_host.h |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/linux/kvm_host.h b/include/linux/kvm_host.h
index 6d58beb65454..5873ac4b9aef 100644
--- a/include/linux/kvm_host.h
+++ b/include/linux/kvm_host.h
@@ -409,7 +409,7 @@ struct kvm_irq_routing_table {
 	 * Array indexed by gsi. Each entry contains list of irq chips
 	 * the gsi is connected to.
 	 */
-	struct hlist_head map[0];
+	struct hlist_head map[];
 };
 #endif
 
